### PR TITLE
Add per layer initialization of  ranges.

### PR DESCRIPTION
**Pull request by asenina**

### CODE REVIEW

**Comment by asenina** [comment]
@vshampor please take a look

File: nncf/quantization/algo.py [hunk 1]
```diff
@@ -912,27 +913,42 @@ def init_range(self):
...
+                raise ValueError(
+                    'Should run range initialization as specified via config,'
+                    'but the initializing data loader is not provided as an extra struct. '
+                    'Refer to `NNCFConfig.register_extra_structs` and the `QuantizationRangeInitArgs` class')
+            data_loader = range_init_args.data_loader
+        
+            self._do_range_init(data_loader, max_num_init_steps, global_init_range_config,
+                                range_init_args.device)
+
+    def get_correct_range_config(self, init_range_config: Dict):
```
Comment (vshampor) [inline nncf/quantization/algo.py:939]: "correct" is too broad - should be more specific with what kind of config this function returns.

File: nncf/quantization/algo.py [hunk 2]
```diff
@@ -1043,13 +1059,35 @@ def get_bit_stats(self):
...
+        else:
+            module_init_range_config = None
+            matched_init_range_config = []
+            for range_init_subconfig in global_init_range_config:
+                target_scopes = range_init_subconfig.get("target_scopes", None)
+                ignored_scopes = range_init_subconfig.get("ignored_scopes", None)
+                target_quantizer_group = range_init_subconfig.get("target_quantizer_group", quantizer_group)
+                if quantizer_group == target_quantizer_group and \
+                    (target_scopes is not None and in_scope_list(str(scope), target_scopes)  or\
+                    ignored_scopes is not None and not in_scope_list(str(scope), ignored_scopes)):
```
Comment (vshampor) [inline nncf/quantization/algo.py:1074]: In quantization builder, we have a "_should_consider_scope..." function that wraps this kind of logic. Do the same here - perhaps make a helper function that will be used both inside "_should_consider_scope..." and here, but taking different `ignored_scopes` and `target_scopes` arguments at runtime.

File: nncf/quantization/algo.py [hunk 3]
```diff
@@ -1043,13 +1059,35 @@ def get_bit_stats(self):
...
+            module_init_range_config = matched_init_range_config[0]
+
         for overridden_scope in scope_overrides.keys():
             if in_scope_list(str(scope), overridden_scope):
                 override_config = scope_overrides[overridden_scope].get('initializer', {}).get("range")
                 if override_config is not None:
                     module_init_range_config = override_config
+
+        if module_init_range_config is None:
+            raise ValueError("Range initialization config is undefined for scope: {}. ".format(str(scope)))
```
Comment (vshampor) [inline nncf/quantization/algo.py:1089]: Will the logic fail here if we specify only per-layer initializations (no global init config), but we don't do it for all present quantizer scopes?

File: tests/quantization/test_range_init.py [hunk 4]
```diff
@@ -243,6 +244,157 @@ def test_scope_overrides(self, wrap_dataloader):
...
+
+    def test_per_layer_range_init_with_correct_possible_config(self, wrap_dataloader, per_layer_range_init_test_struct):
+        config = create_config()
+        config['compression']['initializer']['range'] = per_layer_range_init_test_struct.range_init_config
+        data_loader = self.create_dataloader(wrap_dataloader, config)
+        config.register_extra_structs([QuantizationRangeInitArgs(data_loader)])
+        ctrl, _ = self.create_algo_and_compressed_model(config)
+        for str_scope, range_init_config in per_layer_range_init_test_struct.expected_modules_to_init.items():
+            assert ctrl.modules_to_range_init[str_scope][1] == range_init_config
+
```
Comment (vshampor) [inline tests/quantization/test_range_init.py:398]: Should add a test that checks that the initializers are actually called the required number of times. Use "mock" functionality of pytest for this.

File: nncf/config_schema.py [hunk 5]
```diff
@@ -174,31 +174,56 @@ def with_attributes(schema: Dict, **kwargs) -> Dict:
...
+                                                        "input value histograms to be set as the initial "
+                                                        "value for minimum quantizer input"),
+        "max_percentile": with_attributes(_NUMBER,
+                                            description="For 'percentile' type - specify the percentile of "
+                                                        "input value histograms to be set as the initial "
+                                                        "value for maximum quantizer input"),
+    },
+    "additionalProperties": False,
+}
+EXTENDED_RANGE_INIT_CONFIG_PROPERTIES = {
```
Comment (vshampor) [inline nncf/config_schema.py:199]: This is only "extended" with a per-layer specification functionality. Should call it `PER_LAYER_RANGE_INIT_CONFIG_PROPERTIES`, perhaps.

**Review by vshampor — CHANGES_REQUESTED** [review]

File: nncf/quantization/algo.py (hunk 1 above)
Comment (asenina) [reply nncf/quantization/algo.py]: Done.

File: nncf/quantization/algo.py (hunk 2 above)
Comment (asenina) [reply nncf/quantization/algo.py:1074]: Done.

File: tests/quantization/test_range_init.py (hunk 4 above)
Comment (asenina) [reply tests/quantization/test_range_init.py:399]: Done.

**Comment by asenina** [comment]
Jenkins please retry a build

File: nncf/config_schema.py (hunk 5 above)
Comment (asenina) [reply nncf/config_schema.py]: Done.

**Comment by asenina** [comment]
@vshampor  please merge it

**Review by vshampor — APPROVED** [review]

**Pull request merged**